src/Raven.Server/Documents/Queries/Results/TimeSeriesRetriever.cs

```diff
@@ -686,20 +676,55 @@ string GetSourceAndId()
                 else
                 {
                     if (index == -1 || index == declaredFunction.Parameters.Count) // not found
-                        throw new InvalidQueryException($"Unable to parse TimeSeries name from expression '{(FieldExpression)timeSeriesFunction.Between.Source}'. " +
+                        throw new InvalidQueryException($"Unable to parse TimeSeries name from expression '{timeSeriesFunction.Source}'. " +
                                                             $"'{compound[0]}' is unknown, and no matching argument was provided to time series function '{declaredFunction.Name}'.");
 
                     if (!(args[index] is Document document))
-                        throw new InvalidQueryException($"Unable to parse TimeSeries name from expression '{(FieldExpression)timeSeriesFunction.Between.Source}'. " +
+                        throw new InvalidQueryException($"Unable to parse TimeSeries name from expression '{timeSeriesFunction.Source}'. " +
                                                             $"Expected argument '{compound[0]}' to be a Document instance, but got '{args[index].GetType()}'");
 
                     documentId = document.Id;
                 }
 
-                return ((FieldExpression)timeSeriesFunction.Between.Source).FieldValueWithoutAlias;
+                return timeSeriesFunction.Source.FieldValueWithoutAlias;
             }
         }
 
+        private (DateTime Min, DateTime Max) GetMinAndMax(DeclaredFunction declaredFunction, string documentId, object[] args, TimeSeriesFunction timeSeriesFunction, string source)
+        {
+            DateTime min, max;
+            if (timeSeriesFunction.Last == null)
+            {
+                min = GetDateValue(timeSeriesFunction.Between?.MinExpression, declaredFunction, args) ?? DateTime.MinValue;
+                max = GetDateValue(timeSeriesFunction.Between?.MaxExpression, declaredFunction, args) ?? DateTime.MaxValue;
+            }
+            else
+            {
+                var last = timeSeriesFunction.Last.GetValue(_queryParameters)?.ToString();
+                if (last == null)
+                {
+                    throw new InvalidQueryException($"Time series function '{declaredFunction.Name}' on document '{documentId}' " +
+                                                    $"was unable to read 'Last' expression '{timeSeriesFunction.Last}'");
+                }
+
+                var stats = _context.DocumentDatabase.DocumentsStorage.TimeSeriesStorage.Stats.GetStats(_context, documentId, source);
+                max = stats.End == default
+                    ? DateTime.MaxValue
+                    : stats.End;
+
+                var timeFromLast = RangeGroup.ParseLastFromString(last);
+                min = max.Add(-timeFromLast);
+            }
+
+            if (timeSeriesFunction.Offset.HasValue)
```

Comment:
Not sure if this is correct when we use last, we might skip results, no?

Reply:
added test, seems to work fine :)